### PR TITLE
Add breadcrumb navigation

### LINKED ISSUE #2097: Include bread crumb navigation
### Is your feature request related to a problem?

Currently, navigation in Artemis is not as easy as it could be. Some pages have a back button, others do not.

### Describe the solution you'd like

Bread crumb navigation seems to be a good solution. It shows users where they are and enables easy navigation. If it is put into the top bar it can be present on all Artemis pages.

### Describe alternatives you've considered

Back buttons give less information, are harder to implement on all pages and mostly duplicate the built in browser _back_ feature.
Having no GUI elements for navigation makes it hard for users to know where they are and allows fewer navigation options.

### Additional context

This feature was already discussed in #2083.

### PR DESCRIPTION
**Pull request by FrankeLukas**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I added multiple screenshots/screencasts of my UI changes

### Motivation and Context
See #2097.

### Description
Added bread crumbs to the navigation bar in the top. This is currently restricted to users above the "normal" user (tutors, instructors, admins) and to the routes `<base-url>/#/admin` and `<base-url>/#/course-management`. Better labelling (e.g. a course name instead of the id) will be added in a future PR.

### Steps for Testing

1. Open pages matching the routes `<base-url>/#/admin` or `<base-url>/#/course-management` and check the crumbs are correctly displayed and work
2. Check that users without extra permissions don't get crumbs

Note: Only test that the links work the same as if you typed it in the browser. Not all crumbs make sense right now since some of them are redirects. (Especially the `view` pages need to go. That is part of the reason we don't show it to normal users yet.)

### Screenshots
Examples:
grafik
grafik

### CODE REVIEW

File: src/main/webapp/app/shared/layouts/navbar/navbar.component.ts [hunk 1]
```diff
@@ -95,6 +99,19 @@ export class NavbarComponent implements OnInit, OnDestroy {
         }
     }
 
+    buildBreadcrumbs(fullURI: string) {
+        this.breadcrumbs = [];
+        let path = '';
+        const parts = fullURI.split('/');
+        for (let i = 0; i < parts.length; i++) {
```
Comment (krusche) [inline src/main/webapp/app/shared/layouts/navbar/navbar.component.ts:106]: can we use a modern for loop here?

**Comment by krusche** [comment]
Can you only activate this for routes in `<base-url>/#/admin` and `<base-url>/#/course-management`? So that we can test it in more detail and then activate it for `<base-url>/#/courses` later?

**Comment by FrankeLukas** [comment]
Updated with https://github.com/ls1intum/Artemis/pull/2283/commits/6730592f284a60db98fbeb4aff186346a4f70369.

**Review by MaisiKoleni — APPROVED** [review]
Works well within the current limitation of our routes. This is unfortunately the biggest drawback. Can we improve our routes here and add some alternatives/alias/redirects?

Example problems:
- We can go from the scores (or participations) of an exercise to the CM exercise overview (click `exercises`), but we cannot view the exercise itself (click `1890`).
  image
- For grading (and edit) I can view the exercise (click `1879`), but I cannot go to the CM exercise overview (click `1439`).
  image

Funny is, that I can click on e.g. `quiz-exercises` and get an overview of all quiz exercises in a course. This also works for modelling, text and file-upload. Just not for programming exercises.

In general, as instructor, with our current routes what works and what not is confusing. The exam breadcrumbs are among the useful ones, they work most of the time (concerning what the user has in mind). This is a good first step and well-placed in Artemis, but fixes #2097 only partially.

If I understand the PR correctly this is out of scope, so we definitely need a follow-up here to improve routing and properly fix #2097.

**Comment by FrankeLukas** [comment]
> If I understand the PR correctly this is out of scope, so we definitely need a follow-up here to improve routing and properly fix #2097.

Yes, I started that with #2309 now.

**Review by kloessst — APPROVED** [review]
Works fine for the scope of this PR. Code looks good to me.

File: src/main/webapp/app/shared/layouts/navbar/navbar.component.ts [hunk 2]
```diff
@@ -95,6 +99,24 @@ export class NavbarComponent implements OnInit, OnDestroy {
...
+        let path = '';
+        let index = 0;
+        const parts = fullURI.split('/');
+        for (const part of parts) {
+            const crumb = new Breadcrumb();
+            crumb.label = part;
+            path += part + '/';
+            crumb.uri = path;
+            this.breadcrumbs[index++] = crumb;
+        }
```
Comment (fde312) [inline src/main/webapp/app/shared/layouts/navbar/navbar.component.ts:117]: I think you can change this also to something like this:

```suggestion
        this.breadcrumbs: Breadcrumb[] = parts.map(part => {
    path += part + '/';
    return {label: part, uri: path}
})
```

**Review by fde312 — APPROVED** [review]
Worked for me. Good job! 🚀
I just left you a comment about using `map` instead of the `for` loop, but the for-loop is also okay for me :)

**Review by balazs-czopf — APPROVED** [review]
Tested it on TS1, worked as expected. Nice job!

**Review by philippbauch — APPROVED** [review]
Works with some limitations already mentioned above. ✅ 

Would also like to add to @MaisiKoleni's comment:

Screenshot 2020-11-05 at 22 45 04

It might be a little confusing if different sub-routes (`/scores` and `/grading` in this case) originate from either `/exercises` or `/programming-exercises`.

`/programming-exercises` will navigate to a page otherwise unreachable, so I'd prefer to always use the generic `/exercises` to navigate back to the exercise overview. ✌️

**Comment by krusche** [comment]
> Works with some limitations already mentioned above. ✅
> 
> Would also like to add to @MaisiKoleni's comment:
> 
> Screenshot 2020-11-05 at 22 45 04
> 
> It might be a little confusing if different sub-routes (`/scores` and `/grading` in this case) originate from either `/exercises` or `/programming-exercises`.
> 
> `/programming-exercises` will navigate to a page otherwise unreachable, so I'd prefer to always use the generic `/exercises` to navigate back to the exercise overview. ✌️

These are good improvements for a follow-up

**Review by krusche — APPROVED** [review]

**Pull request merged**